tango/step_caches/local_step_cache.py

```diff
@@ -30,18 +33,37 @@ class LocalStepCache(StepCache):
     """
 
     LRU_CACHE_MAX_SIZE = 8
+    METADATA_FILE_NAME = "cache-metadata.json"
 
     def __init__(self, dir: PathOrStr):
         self.dir = Path(dir)
         self.dir.mkdir(parents=True, exist_ok=True)
 
         # We keep an in-memory cache as well so we don't have to de-serialize stuff
         # we happen to have in memory already.
-        self.weak_cache: MutableMapping[str, Any] = weakref.WeakValueDictionary()
-
+        self.weak_cache: MutableMapping[str, Any]
         # Not all Python objects can be referenced weakly, and even if they can they
         # might get removed too quickly, so we also keep an LRU cache.
-        self.strong_cache: OrderedDict[str, Any] = collections.OrderedDict()
+        self.strong_cache: OrderedDict[str, Any]
+        self._init_mem_caches()
+
+    def _init_mem_caches(self):
+        self.weak_cache = weakref.WeakValueDictionary()
+        self.strong_cache = collections.OrderedDict()
+
+    def __getstate__(self):
+        """
+        We override `__getstate__()` to customize how instances of this class are pickled
+        since we don't want to persist values in the weak and strong in-memory caches
+        during pickling. And WeakValueDictionary can't be pickled anyway.
+        """
+        out = {k: v for k, v in self.__dict__.items() if k not in {"weak_cache", "strong_cache"}}
+        return out
```

Comment:
Unfortunately yes. For example, during distributed training we pass the workspace to worker processes: https://github.com/allenai/tango/pull/232/files#diff-72a1daf1ee839673aee1e5e2b0b2a65111e0e72bbaeaa07a5321ee88882156caR315

Reply:
I think it is in the nature of this class that the only thing we actually want to store is the directory that contains the cache, right? Because the cache guarantees that it will look the same from every process that uses that directory, by definition the directory must be enough to pickle the cache. So I'm thinking, let's just return the directory from `__getstate__`.